src/lazy/text/buffer.rs

```diff
@@ -3046,6 +3057,22 @@ mod tests {
         ]
     }
 
+    matcher_tests_with_macro! {
+        allow_omitting_trailing_optionals
+        match_e_expression
+        "(macro foo (a b+ c? d*) null)"
+        expect_match: [
+            "(:foo 1 2)",
+            "(:foo 1 2 3)",
+            "(:foo 1 2 3 4)",
+            "(:foo 1 2 3 4 5 6)", // implicit rest
+        ],
+        expect_mismatch: [
+            "(:foo 1)",
+            "(:foo)",
```

Comment:
```suggestion
            "(:foo)",
            "(:foo 1 2 3 4 (: 5) 6)", // final argument must be rest or expression group, but not a mix of both
```
...unless this is already covered elsewhere.